diff --git a/Makefile b/Makefile
index 90f84f9..2aa7c90 100644
--- a/Makefile
+++ b/Makefile
@@ -1515,7 +1515,8 @@ clean := -f $(if $(KBUILD_SRC),$(srctree)/)scripts/Makefile.clean obj
 endif	# skip-makefile
 
 PHONY += FORCE
-FORCE: drivers/net/cns3xxx-nas/cns3xxx-nas/cns3xxx-nas-patch
+FORCE:
+#FORCE: drivers/net/cns3xxx-nas/cns3xxx-nas/cns3xxx-nas-patch
 
 # Declare the contents of the .PHONY variable as phony.  We keep that
 # information in a variable so we can use it in if_changed and friends.
